### PR TITLE
add example how to use ParILU to generate an ILU-preconditioned solver

### PR DESCRIPTION
**Pull request by hartwiganzt**
Addition of an example how to use the ParILU algorithm to generate an ILU preconditioner and use it in an iterative solver.

### CODE REVIEW

File: examples/ilu-preconditioned-solver/doc/intro.dox [hunk 1]
```diff
@@ -0,0 +1,5 @@
+<a name="Intro"></a>
+<h1>Introduction</h1>
+
+<h3> About the example </h3>
```
Comment (thoasm) [inline examples/ilu-preconditioned-solver/doc/intro.dox]: Should we leave the description / documentation for this example empty?

File: examples/ilu-preconditioned-solver/ilu-preconditioned-solver.cpp [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+    auto ilu_pre_factory =
+        gko::preconditioner::Ilu<gko::solver::LowerTrs<>,
+                                 gko::solver::UpperTrs<>, false>::build()
+            .on(exec);
+
+    // Use incomplete factors to generate ILU preconditioner
+    auto ilu_preconditioner = ilu_pre_factory->generate(gko::share(par_ilu));
+
+    // Use preconditioner inside GMRES solver factory
+    // Generating a factory makes sense if there are several systems for which
```
Comment (thoasm) [inline examples/ilu-preconditioned-solver/ilu-preconditioned-solver.cpp]: Why does generating a "factory" make sense when you could reuse an ILU preconditioner?
Did you mean "using a generated preconditioner in the factory makes sense if several systems are expected to be effective with it" or similar?

Comment (hartwiganzt) [reply examples/ilu-preconditioned-solver/ilu-preconditioned-solver.cpp]: I tried to answer the question in the updated comment: a prec+solver combination might be effective for several problems, this is why we may generate a factory of this kind.

File: examples/ilu-preconditioned-solver/ilu-preconditioned-solver.cpp [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+    // Print solution
+    std::cout << "Solution (x): \n";
+    write(std::cout, lend(x));
+
+    // Calculate residual
+    auto one = gko::initialize<vec>({1.0}, exec);
+    auto neg_one = gko::initialize<vec>({-1.0}, exec);
+    auto res = gko::initialize<vec>({0.0}, exec);
+    A->apply(lend(one), lend(x), lend(neg_one), lend(b));
+    b->compute_norm2(lend(res));
```
[the remark this answers is not in the document]
Comment (hartwiganzt) [reply examples/ilu-preconditioned-solver/ilu-preconditioned-solver.cpp]: Done.

**Pull request merged**